### PR TITLE
[16.07] install Conda at startup; enable conda_auto_init in sample conf

### PR DESCRIPTION
**Pull request by martenson**
- adjust wording

### CODE REVIEW

**Comment by jmchilton** [comment]
This installs conda when Galaxy starts up - not before the first time a tool dependency attempts to be installed - I'm not sure this is a good idea - I'd prefer to change this default only if it this conda installation were to occur as part of a dependency install. Thoughts?

**Comment by martenson** [comment]
@jmchilton are you referring to the change in init.py or am I misunderstanding how it works?

**Comment by jmchilton** [comment]
Yeah - your change of that default causes this code https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/deps/resolvers/conda.py#L92 to run when Galaxy starts up by default. I'd prefer Galaxy never access the Internet for instance unless explicitly asked too.

**Comment by bgruening** [comment]
My feeling is that during startup admins expect such changes, like database migrations, tool migrations, wheel updates and so on. During tool installation I would not expect a conda installation and any error could be confusing.

File: config/galaxy.ini.sample [hunk 1]
```diff
@@ -164,29 +164,31 @@ paste.app_factory = galaxy.web.buildapp:app_factory
...
+# ordering is to the use the Tool Shed for tools installed that way, use local
+# Galaxy packages, and then use Conda if available.
 # See https://github.com/galaxyproject/galaxy/blob/dev/doc/source/admin/dependency_resolvers.rst
 # for more information on these options.
 #dependency_resolvers_config_file = config/dependency_resolvers_conf.xml
 
-# Following conda dependeny resolution options are experimental as of early
-# 2016. These will change the defaults for each conda resolver, but multiple
-# resolvers can be configured independently and these options overridden in
+# Following Conda dependency resolution options will change the defaults for each Conda resolver,
```
Comment (nsoranzo) [inline config/galaxy.ini.sample:186]: s/Following/The following/

**Comment by jmchilton** [comment]
You lost me at "tool migration" but I guess we do update wheels at startup. I don't think we will when we pip install Galaxy those. @natefoo are you okay with Galaxy automatically install conda when it starts up by default?

**Comment by bgruening** [comment]
We did a lot of `tool migrations` during the last years, replacing tools previously shipped with Galaxy core to TS ones.
If we install Galaxy with pip we can also ship a local miniconda tarball and install this during startup, isn't it?

**Comment by mvdbeek** [comment]
If the consensus is to install conda when the resolver is first used I could do a PR on @martenson's branch with [this](https://github.com/martenson/galaxy/compare/conda_config...mvdbeek:conda_install_first_use?expand=1)

**Comment by martenson** [comment]
@jmchilton 

> automatically install conda when it starts up by default?

It is not just start up. You have to major-upgrade (release switch) your Galaxy and then start up. 
And currently for updating wheels you pull latest and then it forces you to download wheels.
And for first ever startup you have to fetch dozens of wheels and there is no easy workaround.

**Comment by martenson** [comment]
If we delay the Conda installation attempt until the resolver is first used we will also force admins to try installing a (conda-only-requirement) tool to make sure Conda installs fine. Wouldn't they benefit from this information during startup when they are setting things up? Surely the Conda installation can fail under some circumstances...

**Comment by mvdbeek** [comment]
> Wouldn't they benefit from this information during startup when they are setting things up?

That is true ... I think right now it probably makes sense to install conda at startup. IMO we can revisit this decision once `pip install galaxy` also fetches wheels and the conda install slows things down disproportionately.

**Comment by jmchilton** [comment]
I'm -0 on this - I think this is making the assumption that conda will be used - it should be considered optional and the assumption shouldn't be that it will be used automatically. There are enough +1s that this will be merged anyway and I'm okay with that - I see the upsides.

**Comment by martenson** [comment]
@jmchilton Do we want to motivate deployers to move towards Conda? Do we want to troubleshoot less failing TS package recipes installations? Do we want to test this on variety of deployments?

edit: These I see as benefits of merging this. Don't we ultimately want to have Conda as the first entry in the dependency_resolver.conf?

**Comment by bgruening** [comment]
Traditional Tool dependencies are also installed by default and we want to enable reproducible workflows, which is only possible if the dependencies are also tracked and reproducible on other systems. We agreed to go with conda so we should activate it on all levels by default. If someone wants to use modules or any other manual system, it is easy enough to deactivate these settings.

Make it easy and reproducible by default! :+1:

**Comment by mvdbeek** [comment]
The startup test on OSX probably fails because it's timing out during the conda install ... or because the prefix may be ridiculously long on travis. @martenson can you include https://github.com/martenson/galaxy/pull/15 ?

**Comment by natefoo** [comment]
I am +1 on installing conda at startup time, and I agree that this should be done by default now, since it is our preferred dependency resolver going forward, and TS build recipes will no longer be required, even for IUC/devteam tools. TS tools will eventually stop working if we don't make this the default.

I also prefer to have it done at startup time rather than "some time in the future when Galaxy is running"

**Comment by martenson** [comment]
PR from @mvdbeek was added to adjust the starting wait (which was failing on OSX travis)

**Comment by mvdbeek** [comment]
👍 once the tests pass.

**Comment by bgruening** [comment]
Green, Green. Thanks to all!

**Pull request merged**